### PR TITLE
feat: Checkbox group should have a select all option

### LINKED ISSUE #7275: [Feature] Checkbox group should have a select all option
## Summary

Checkbox group should provide an option for user to add a select all option in the property itself

## Motivation

In a scenario there might be a large number of option to select in the checkbox group, so user should have an option to select all instead of selecting it one by one. A property can be implemented in the checkbox group itself which on set true can add a select all checkbox at the top of the checkbox group to select all checkboxes in the group

## UI Implementation suggestion

Screenshot 2021-09-09 at 10 27 59 AM

## Link to Designs
https://www.figma.com/file/w9CaXHF5iYcrLtiEbAHcFu/Original-Theme-Widgets-Library?node-id=1015%3A7726

### PR DESCRIPTION
**Pull request by wmdev0808**
## Description

Checkbox group should provide an option for user to add a select all option in the property itself

In a scenario there might be a large number of option to select in the checkbox group, so user should have an option to select all instead of selecting it one by one. A property can be implemented in the checkbox group itself which on set true can add a select all checkbox at the top of the checkbox group to select all checkboxes in the group

Fixes #7275 

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

- Test A
- Test B

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:white_circle: Total coverage has not changed</summary>

    // Code coverage diff between base branch:release and head branch: feat/7275-checkboxgroup-add-select-all 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 55.14 **(0)** | 36.64 **(-0.01)** | 34.58 **(0.01)** | 55.67 **(0)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :sparkles: :new: | **app/client/src/widgets/CheckboxGroupWidget/constants.ts** | **100** | **100** | **100** | **100**
 :red_circle: | app/client/src/widgets/CheckboxGroupWidget/component/index.tsx | 29.03 **(4.03)** | 5.56 **(-6.94)** | 0 **(0)** | 36 **(2.67)**
 :red_circle: | app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx | 51.35 **(-7.38)** | 26.92 **(-2.25)** | 38.1 **(-6.34)** | 52.17 **(-7.15)**
 :green_circle: | app/client/src/widgets/CheckboxWidget/component/index.tsx | 96.97 **(0.42)** | 67.74 **(3.74)** | 90.91 **(0.91)** | 96.15 **(0.32)**</details>

### CODE REVIEW

**Review by somangshu — CHANGES_REQUESTED** [review]
LGTM, I have tested the Deploy preview and it works well, Left a couple of comments here.
- When the space is less the select all option collides with the option underneath 
 Screenshot 2021-10-25 at 4 16 09 PM
- ~~The options should go below disabled~~ 
 Screenshot 2021-10-25 at 4 26 26 PM
- ~~The select all option and not selected checkbox color should be `grey-9`~~
- ~~The selected option colour should be `grey-10`~~
 

Documentation change will be required

**Comment by wmdev0808** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1400349497

**Review by somangshu — DISMISSED** [review]

**Comment by somangshu** [comment]
@wmdev0808 The solution to first issue is not apt, The additional gap below the select option seem to be mis-aligned in the inline more 
Screenshot 2021-11-02 at 11 49 20 AM

Also the cypress test have failed due to the relevant change you have made, Kindly check

**Comment by wmdev0808** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1411916481

**Comment by somangshu** [comment]
> * When the space is less the select all option collides with the option underneath
>   Screenshot 2021-10-25 at 4 16 09 PM

@wmdev0808 the previous issue has resurfaced because of the change, Kindly check. 

My analysis of the problem.
- When inline is true, there is additional space underneath 
Screenshot 2021-11-10 at 2 40 41 PM
- When inline is false, the space is gone because of the recent change and this way the text collides 
Screenshot 2021-11-10 at 2 41 41 PM
- The solution could be to persist the select all option in one line, Let it overflow.

**Review by somangshu — CHANGES_REQUESTED** [review]
As per the above comment

**Comment by wmdev0808** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1443487133

File: app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx [hunk 1]
```diff
@@ -163,6 +151,30 @@ class CheckboxGroupWidget extends BaseWidget<
...
+            isTriggerProperty: false,
+            validation: {
+              type: ValidationTypes.BOOLEAN,
+            },
+          },
+          {
+            propertyName: "isSelectAll",
+            label: "Select All Options",
+            controlType: "SWITCH",
+            helpText: "Controls whether select all control is shown",
```
Comment (somangshu) [inline app/client/src/widgets/CheckboxGroupWidget/widget/index.tsx:170]: @wmdev0808 can you make a small change here > `Controls whether select all option is shown`

**Review by somangshu — DISMISSED** [review]
Change LGTM

**Review by somangshu — APPROVED** [review]

**Comment by wmdev0808** [comment]
https://github.com/appsmithorg/appsmith/actions/runs/1443627099

**Comment by somangshu** [comment]
@wmdev0808 please update this with the latest release! Check other PR too

**Comment by somangshu** [comment]
@wmdev0808 the tests have passed, You can close this is you have the QA sign off

**Review by somangshu — APPROVED** [review]

**Pull request merged**